tools/stats/monitor.py

```diff
@@ -89,49 +161,118 @@ def __init__(
+
+    def _collect_data(self) -> None:
         """
-        runs the main loop of the program.
-        the first json record is the metadata of the run,
-        including the start time, end time, and the interval of the log.
+        Collects the data every data_collect_interval (in seconds).
         """
+        while not self.exit_event.is_set():
+            try:
+                # collect cpu, memory and gpu metrics
+                memory = psutil.virtual_memory().percent
+                cpu_percent = psutil.cpu_percent()
+                processes = self._get_process_info()
+                gpuList = self._collect_gpu_data()
+
+                data = UsageData(
+                    cpu_percent=cpu_percent,
+                    memory_percent=memory,
+                    processes=processes,
+                    gpu_list=gpuList,
+                )
+                if self._debug_mode:
+                    print(f"collecting data {data}")
 
+                self.shared_resource.add_data(data)
+            except Exception as e:
+                if self._debug_mode:
+                    print(f"error detected: {str(e)}")
+                self.shared_resource.add_error(e)
+            finally:
+                time.sleep(self._data_collect_interval)
+
+    def _output_data(self) -> None:
+        """
+        output the data.
+        """
         self._summary_info["start_time"] = datetime.datetime.now().timestamp()
         self.log_json(self._summary_info)
 
-        # start data collection
-        while not self._kill_now:
+        while not self.exit_event.is_set():
             collecting_start_time = time.time()
             stats = {}
             try:
+                data_list, error_list = self.shared_resource.get_and_reset()
+                if self._debug_mode:
+                    print(
+                        f"collected data: {len(data_list)}, errors found: {len(error_list)}"
+                    )
+                # records and clears found errors
+                errors = list(set(error_list))
+
+                # if has errors but data list is None, a bug may exist in the monitor code, log the errors
+                if not data_list and len(errors) > 0:
+                    raise ValueError(
+                        f"no data is collected but detected errors during the interval: {errors}"
+                    )
+                if not data_list:
+                    # pass since no data is collected
+                    continue
+
                 stats.update(
                     {
                         "level": "record",
                         "time": datetime.datetime.now().timestamp(),
                     }
                 )
+
                 # collect cpu and memory metrics
-                memory = psutil.virtual_memory()
-                used_cpu_percent = psutil.cpu_percent()
+                total_cpu = sum(usageData.cpu_percent for usageData in data_list)
```

Comment:
It seems like a good idea to more this block to compute `avg_*` and `max_*` into a separate function, i.e. `generate_stats` for example.  Then you could just use this to compute both cpu, memory, and other metrics in the same way.

The pseudo code looks like this:

```
cpu_stats = generate_stats(data_list, "cpu")
memory_stats = generate_stats(data_list, "memory")
...
stats.update(
    {
        "cpu": cpu_stats,
        "memory": memory_stats,
        ...
    }
)
```